doc/api18970.md

```diff
@@ -0,0 +1,408 @@
+cooperative threads have a priority higher than any preemptible thread,
+and meta-IRQ threads have a priority higher than any other cooperative thread.
+
+Most Zephyr threads are cooperative, and some code may take advantage of
+this by assuming the behavior of reschedule points reached during an
+operation will be guided by the invoking thread being cooperative.
+
+Given this, define the following terms:
+
+* A thread **suspends** when it voluntarily invokes a function that
+  causes it to transition from *Running* to *Suspended*.
+  (`k_sleep()` is such a function.)
+* A thread **waits** when it voluntarily invokes a function that causes
+  it to transition from *Running* to *Waiting*.  (`k_poll()` is such a
+  function.)
+* A thread **yields** when it voluntarily invokes a function that causes
+  it to transition from *Running* to *Ready*.  (`k_yield()` is such a
+  function.)
+* A thread **sleeps** when it voluntarily invokes a function that causes
+  it to suspend, wait, or yield.
+* A thread is **preempted** when it is *Running* but a reschedule point
+  (involuntarily) transitions it to *Ready*.
+
+A **context switch** occurs when the reschedule point selects a thread
+other than the current thread to execute next.
+
+For API behavior we are primarily interested in whether invoking a
+function can directly cause a thread invoking it to sleep.
+
+#### Digresssions and random notes
+
+*Suspended* could be perceived as a variant of *Waiting* where the
+releasing event is an invocation of `k_thread_resume()`, possibly
+performed by the callback of a timeout associated with the thread.
+
+For operations that cause a transition from *Running* to *Waiting* a
+timeout value `K_NO_WAIT` indicates that the operation should fail
+rather than change the thread state.  The failure is indicated with an
+operation-specific error code such as `-EAGAIN`, `-EBUSY`, `-ENOMEM`,
+and `-ENOMSG`.
+
+* `z_pend_curr` transitions the current thread to **Waiting**;
+* `k_yield` transitions the current thread to **Ready** and moves it to
+  the end of the ready queue.  It also explicitly overrides the default
+  preservation of cooperative threads as the current thread.
+* `k_sleep` invokes `k_yield()` if timeout is zero, otherwise suspends
+  the thread until the timeout.
+
+An operation cannot be invoked from an ISR if:
+* It transitions the current thread to *Suspended* (per assert in
+  `z_tick_sleep()`)
+* It transitions the current thread to *Waiting* (per documentation of
+  e.g. `k_sem_take()`)
+* It transitions the current thread to *Ready* (per assert in
+  `k_yield()`)
+
+Generalizing this, and consistent with the informal answer to issue
+21341, an operation cannot be invoked when invoked from a non-thread
+context (interrupt or pre-kernel) if it could cause its invoking thread
+to sleep, because without in invoking thread it cannot implement the
+required behavior.
+
+Kernel API assessed using terminology below, plus "cooperative-safe"
```

Comment:
This table was very helpful. I would not call it a digression or random note, but maybe promote it up higher.